### PR TITLE
ROX-14912 load testing for processes listening on ports

### PR DESCRIPTION
**Pull request by JoukoVirtanen**
## Description

Adds container endpoints to the fake sensor data. These endpoints also have information about the fake processes that "launched" them. The purpose of this is to do load testing of processes listening on ports.

## Checklist
- [x] Investigated and inspected CI test results
- [x] Saw entries in process_listening_on_ports table
- [x] Used to do load testing
- [x] Ran scale tests in CI

## Testing Performed

Load testing results and how they were done can be found here https://docs.google.com/document/d/1Es96w-jqYXxxmWDt8Sekm5VN8iMLDqBIkoPmDHwOvH8/edit#

How to see entries in process_listening_on_ports table

Started load similar to https://docs.engineering.redhat.com/display/StackRox/Long-running+cluster with the largest change being that the cluster was not set to run for a long time

Got the postgres password with 
kubectl -n stackrox get secret central-db-password -o yaml

Login to postgres
ks exec -it central-db-<pod-name> -- psql
Use password obtained above

Use the correct database
\c central_active

Count the number of rows
select count(*) from process_listening_on_ports ;

or see the rows themselves
select * from process_listening_on_ports ;

### CODE REVIEW

File: sensor/kubernetes/fake/flows.go [hunk 1]
```diff
@@ -102,6 +103,18 @@ func (w *WorkloadManager) getRandomHostConnection(ctx context.Context) (manager.
...
+			ProcessExecFilePath:	process.ExecFilePath,
+			ProcessArgs:		process.Args,
+		}
+	}
+
+	return nil
+}
+
 // manageFlows should be called via `go manageFlows` as it will run forever
 func (w *WorkloadManager) manageFlows(ctx context.Context, workload NetworkWorkload) {
```
Comment (JoukoVirtanen) [inline sensor/kubernetes/fake/flows.go:119]: Refactor to split up into smaller functions

File: sensor/kubernetes/fake/deployment.go [hunk 2]
```diff
@@ -6,18 +6,59 @@ import (
...
+	lock		sync.RWMutex
+}
+
+func newProcessPool() *ProcessPool {
+	return &ProcessPool{
+		ProcessPool:	make(map[string][]*storage.ProcessSignal),
+	}
+}
+
+func (p *ProcessPool) add(val *storage.ProcessSignal) {
```
Comment (JoukoVirtanen) [inline sensor/kubernetes/fake/deployment.go:38]: Locks might not be needed in this function and the remove function, as I don't think there will be multiple simultaneous calls with the same containerId.

**Comment by porridge** [comment]
I'm afraid I'm not the best person to review this code, I only happened to change a single line there in a mass refactoring.

**Comment by JoukoVirtanen** [comment]
> I'm afraid I'm not the best person to review this code, I only happened to change a single line there in a mass refactoring.

Sorry. I requested you since you were a recommended reviewer. I understand.

File: sensor/kubernetes/fake/deployment.go [hunk 3]
```diff
@@ -6,18 +6,63 @@ import (
...
+	return &ProcessPool{
+		ProcessPool: make(map[string][]*storage.ProcessSignal),
+	}
+}
+
+func (p *ProcessPool) add(val *storage.ProcessSignal) {
+	p.lock.Lock()
+	defer p.lock.Unlock()
+
+	p.ProcessPool[val.ContainerId] = append(p.ProcessPool[val.ContainerId], val)
```
Comment (JoukoVirtanen) [inline sensor/kubernetes/fake/deployment.go:43]: Place a cap on how many processes there can be. Replace old processes when cap is reached.

**Review by Molter73 — COMMENTED** [review]
Overall looks good! Left a few comments/questions.

File: sensor/kubernetes/fake/deployment.go [hunk 4]
```diff
@@ -6,18 +6,84 @@ import (
...
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 )
 
+var (
+	processPool = newProcessPool()
+)
+
+// ProcessPool stores processes by containerID using a map
+type ProcessPool struct {
+	ProcessPool map[string][]*storage.ProcessSignal
```
[the remark this answers is not in the document]
Comment (JoukoVirtanen) [reply sensor/kubernetes/fake/deployment.go:29]: Sure. In flows.go there is pool struct that has a pool field. I was just going along with what was there, but I agree with your point.

File: sensor/kubernetes/fake/deployment.go [hunk 5]
```diff
@@ -6,18 +6,84 @@ import (
...
+	}
+
+	return size
+}
+
+func (p *ProcessPool) add(val *storage.ProcessSignal) {
+	p.lock.Lock()
+	defer p.lock.Unlock()
+
+	capacity := 10000
```
[the remark this answers is not in the document]
Comment (JoukoVirtanen) [reply sensor/kubernetes/fake/deployment.go:53]: Moved it to ProcessPool.

File: sensor/kubernetes/fake/flows.go [hunk 6]
```diff
@@ -102,6 +103,94 @@ func (w *WorkloadManager) getRandomHostConnection(ctx context.Context) (manager.
...
+		}
+
+		containerID, ok := containerPool.randomElem()
+		if !ok {
+			log.Error("found no containers in pool")
+			continue
+		}
+
+		closeTS, err := types.TimestampProto(time.Now().Add(-5 * time.Second))
+		utils.CrashOnError(err)
```
[the remark this answers is not in the document]
Comment (JoukoVirtanen) [reply sensor/kubernetes/fake/flows.go:175]: I refactored this part. I did not actually write that so I don't know why CrashOnError was used here. I changed it to a log statement.

**Review by Molter73 — APPROVED** [review]
LGTM!

File: sensor/kubernetes/fake/deployment.go [hunk 7]
```diff
@@ -6,18 +6,84 @@ import (
...
+
+func newProcessPool() *ProcessPool {
+	return &ProcessPool{
+		Processes: make(map[string][]*storage.ProcessSignal),
+		Capacity:  10000,
+	}
+}
+
+func (p *ProcessPool) getProcessPoolSize() int {
+	size := 0
```
Comment (connorgorman) [inline sensor/kubernetes/fake/deployment.go:42]: is it possible to just keep track of this in a variable instead of computing it?

Comment (JoukoVirtanen) [reply sensor/kubernetes/fake/deployment.go]: Done

**Pull request merged**